src/store/index-level.ts

```diff
@@ -1,213 +1,393 @@
+ * A LevelDB implementation for indexing the messages and events stored in the DWN.
  */
 export class IndexLevel {
-  config: IndexLevelConfig;
-
   db: LevelWrapper<string>;
+  config: IndexLevelConfig;
 
   constructor(config: IndexLevelConfig) {
     this.config = {
       createLevelDatabase,
-      ...config
+      ...config,
     };
 
-    this.db = new LevelWrapper<string>({ ...this.config, valueEncoding: 'utf8' });
+    this.db = new LevelWrapper<string>({
+      location            : this.config.location!,
+      createLevelDatabase : this.config.createLevelDatabase,
+      keyEncoding         : 'utf8'
+    });
   }
 
   async open(): Promise<void> {
-    return this.db.open();
+    await this.db.open();
   }
 
   async close(): Promise<void> {
-    return this.db.close();
+    await this.db.close();
+  }
+
+  /**
+ * deletes everything in the underlying index db.
+ */
+  async clear(): Promise<void> {
+    await this.db.clear();
   }
 
   /**
-   * Adds indexes for a specific data/object/content.
-   * @param dataId ID of the data/object/content being indexed.
+   * Put an item into the index using information that will allow it to be queried for.
+   *
+   * @param tenant
+   * @param itemId a unique ID that represents the item being indexed, this is also used as the cursor value in a query.
+   * @param indexes - (key-value pairs) to be included as part of indexing this item. Must include at least one indexing property.
+   * @param options IndexLevelOptions that include an AbortSignal.
    */
   async put(
     tenant: string,
-    dataId: string,
-    indexes: { [property: string]: unknown },
+    itemId: string,
+    indexes: KeyValues,
     options?: IndexLevelOptions
   ): Promise<void> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
-    indexes = flatten(indexes);
 
-    const operations: LevelWrapperBatchOperation<string>[] = [ ];
+    // ensure we have something valid to index, remove any empty strings from index properties
+    removeEmptyStrings(indexes);
+    if (isEmptyObject(indexes)) {
+      throw new DwnError(DwnErrorCode.IndexMissingIndexableProperty, 'Index must include at least one valid indexable property');
+    }
 
-    // create an index entry for each property in the `indexes`
-    for (const propertyName in indexes) {
-      const propertyValue = indexes[propertyName];
+    const tenantPartition = await this.db.partition(tenant);
+    const indexOps: LevelWrapperBatchOperation<string>[] = [];
 
-      // NOTE: appending data ID after (property + value) serves two purposes:
-      // 1. creates a unique entry of the property-value pair per data/object
-      // 2. when we need to delete all indexes of a given data ID (`delete()`), we can reconstruct the index keys and remove the indexes efficiently
+    // create an index entry for each property
+    // these indexes are all sortable lexicographically.
+    for (const sortProperty in indexes) {
```

Comment:
Nit: can we rename `sortProperty` -> `indexName`, and `sortValue` -> `indexValue`? 'coz the word "sort" is implied for an index.